### PR TITLE
Add complete formatting infrastructure for formatting HIT files

### LINKED ISSUE #10845: HIT: Formatter
## Rationale

A HIT formatter that can read a file and format it so that it has consistent spacing would be ideal for keeping input files and test specifications consistent throughout the project.

## Description

We need a tool that can format HIT files like clang-format formats source files. This formatter should support a max characters per line parameter much like clang-format does. This capability will be used in the "tests" specification to keep longer requirement statements from getting too unwieldy. 

## Impact

None - new capability

Tag @rwcarlsen

### PR DESCRIPTION
**Pull request by rwcarlsen**
I'd like to get a few more things in with this - e.g. git hooks/scripts, prepare some corresponding CIVET checks, etc. before we review/merge it.

This adds a CLI flag to MOOSE executables for auto-formatting input
files in a canonical format.  The canonical format can be customized by
writing a config file named 'moosestyle.i' in hit format like this:

```
     [format]
       # these parameters set the correspondingly named Formatter member variables.  See them
       # for detailed descriptions of what they do.
       indent_string = "  "
       line_length = 100
       canonical_section_markers = true

       # This section specifies parameter/section sorgin order as provided by the formatter's
       # addPattern member function. The content for this section mirrors the structure of
       # the hit file being formatted.  Each section name is a regex (limited to valid hit
       # identifier characters). The fields and subsections within each section specify an
       # order; any field values are ignored. See the docs for that function for more details.
       # Sorting is performed depth-first by rules in lexical order.
       [sorting]
         [.*]
           [.*]
             # 'first' fields in doubly-nested section go first. This rule must go first in
             # order to not override any later-specified ordering of higher-level sections
             # that would occur because the higher-level section matchers for this rule
             # would match every section.
             first = FooBar
           []
         []
         [foo]        # section 'foo' goes first (before other sections)
           bar = bla  # field 'bar' (in the foo section) goes first
           ** = bla   # double glob is placeholder for unordered fields/sections
           baz = bla  # field 'baz' goes last
         []
       []
     []
```

The governing 'moosestyle.i' file for formatting any particular input
file is discovered by looking for one that exists starting in the
current working directory and recursively checking parent directories
until one is found.  If none is found, then a default canonical format
is used.  A MooseUtils::discoverMooseStyleFile function finds relevant
'moosestyle.i' files, and a MooseUtils::formatHit function auto-formats
input files.

[edit] This no longer adds any flags to MOOSE.  In order to facilitate CIVET checks and formatting without compiling moose, formatting has been implemented in a standalone binary in the hit directory.  Scripts in moose use that binary (and build it if necessary).  A git plugin, git hook, and formatting scripts have been added as well.

Fixes #10845

### CODE REVIEW

**Comment by rwcarlsen** [comment]
If we want input file formatting as part of the precheck, then we probably need to have an executable separate from the moose app binaries that formats the content.  If we do this, then would we still want formatting capability integrated into moose app binaries?

**Comment by rwcarlsen** [comment]
This is converging on a stable state.  We need to agree on a canonical format before we can do the "big format".  @idaholab/moose-team - suggestions? - mostly:

* how many spaces for indent (keep 2)?
* what sections do we want to sort first vs last and in what order?
* Do we want to force certain field order(s) beyond having `type=...` first?

**Comment by permcody** [comment]
> how many spaces for indent (keep 2)?

Yes 

> what sections do we want to sort first vs last and in what order?

I'd suggest that we create a list (we actually used to have one in an old commit of MOOSE) that handles most of the major sections. Perhaps we resurrect and create a more comprehensive one? Sections that don't appear in the list would appear at the end.

> Do we want to force certain field order(s) beyond having type=... first?

Probably not

**Comment by rwcarlsen** [comment]
The formatter supports sorting stuff into both the beginning and end of sections.  So we don't have to include every section in a sort order in order to have certain sections sort last - e.g. we could specify only that Mesh goes first and Output goes last if we wanted - everything else would be left alone in between them.

**Comment by permcody** [comment]
This looks pretty good. Let's go ahead and check in the file Brian reference and make sure that works, then this can go in.

File: python/peacock/tests/input_tab/InputTreeWriter/gold/simple_diffusion_inactive.i [hunk 1]
```diff
@@ -1,11 +1,11 @@
-inactive = 'Kernels BCs Executioner'
 [Mesh]
   type = GeneratedMesh
   dim = 2
   nx = 10
   ny = 10
 []
 
+inactive = 'Kernels BCs Executioner'
```
Comment (brianmoose) [inline python/peacock/tests/input_tab/InputTreeWriter/gold/simple_diffusion_inactive.i]: Do we really want `inactive` to be somewhere in the middle? It should probably be at the top, right?

Comment (permcody) [reply python/peacock/tests/input_tab/InputTreeWriter/gold/simple_diffusion_inactive.i]: Definitely at the top!

Comment (rwcarlsen) [reply python/peacock/tests/input_tab/InputTreeWriter/gold/simple_diffusion_inactive.i]: That would probably be a good idea - "inactive" should probably always go first.

**Pull request closed without merging**